include/v4l2_camera/hysteresis_state_machine.hpp

```diff
@@ -0,0 +1,167 @@
+#define HYSTERESIS_STATE_MACHINE_HPP_
+
+#include <tracetools/utils.hpp>
+
+#include <diagnostic_msgs/msg/diagnostic_status.hpp>
+
+#include <string>
+#include <variant>
+
+namespace custom_diagnostic_tasks
+{
+using DiagnosticStatus_t = unsigned char;
+
+// Helper struct to express state machine nodes
+struct StateBase
+{
+  StateBase(const DiagnosticStatus_t lv) : level(lv), num_observations(1) {}
+
+  DiagnosticStatus_t level;
+  size_t num_observations;
+};
+
+struct Stale : public StateBase
+{
+  Stale() : StateBase(diagnostic_msgs::msg::DiagnosticStatus::STALE) {}
+};
+
+struct Ok : public StateBase
+{
+  Ok() : StateBase(diagnostic_msgs::msg::DiagnosticStatus::OK) {}
+};
+
+struct Warn : public StateBase
+{
+  Warn() : StateBase(diagnostic_msgs::msg::DiagnosticStatus::WARN) {}
+};
+
+struct Error : public StateBase
+{
+  Error() : StateBase(diagnostic_msgs::msg::DiagnosticStatus::ERROR) {}
+};
+
+using StateHolder = std::variant<Stale, Ok, Warn, Error>;
+
+StateHolder generate_state(const DiagnosticStatus_t & state)
+{
+  switch (state) {
+    case diagnostic_msgs::msg::DiagnosticStatus::STALE:
+      return Stale{};
+    case diagnostic_msgs::msg::DiagnosticStatus::OK:
+      return Ok{};
+    case diagnostic_msgs::msg::DiagnosticStatus::WARN:
+      return Warn{};
+    case diagnostic_msgs::msg::DiagnosticStatus::ERROR:
+      return Error{};
+    default:
+      throw std::runtime_error("Undefined status");
+  }
+}
+
+static std::string get_level_string(DiagnosticStatus_t level) {
+  switch(level) {
+    case diagnostic_msgs::msg::DiagnosticStatus::OK:
+      return "OK";
+    case diagnostic_msgs::msg::DiagnosticStatus::WARN:
+      return "WARN";
+    case diagnostic_msgs::msg::DiagnosticStatus::ERROR:
+      return "ERROR";
+    case diagnostic_msgs::msg::DiagnosticStatus::STALE:
+      return "STALE";
+    default:
+      return "UNDEFINED";
+  }
+}
+
+static DiagnosticStatus_t get_level(const StateHolder & state)
+{
+  return std::visit([](const auto & s) { return s.level; }, state);
+}
+
+static size_t get_num_observations(const StateHolder & state)
+{
+  return std::visit([](const auto & s) { return s.num_observations; }, state);
+}
+
+class HysteresisStateMachine
+ {
+public:
+  /**
+   * \brief Constructs HysteresisStateMachine, which implements a smoothing
+            filter over observation
+   *
+   * \param num_frame_transition The number of the successive observations for
+            the status transition. E.g., the status will not be changed from OK
+to WARN until successive `num_frame_transition` WARNs are observed.
+   * \param immediate_error_report If true, errors will be reported immediately
+            once observed; otherwise, the hysteresis damping method using
+            `num_frame_transition` will be adopted
+   * \param immediate_relax_state if true, reported state will immediately
+            change if better state that the current one is observed
+   */
+  HysteresisStateMachine(
```

Comment:
`Constructors that are callable with a single argument must be marked explicit to avoid unintentional implicit conversions (fix available)clang-tidy[google-explicit-constructor](https://clang.llvm.org/extra/clang-tidy/checks/google/explicit-constructor.html)`
```suggestion
  explicit HysteresisStateMachine(
```